### PR TITLE
KeyCloack | Integration test

### PR DESCRIPTION
**Pull request by naveenpaul1**
### Explain the Changes
1. Integration test for Keycloack
2. Validate assume role policy document iam structure 

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1. node  ./node_modules/.bin/mocha ./src/test/integration_tests/api/sts/test_sts_keycloak_integration.js

make run-single-test testpath=integration_tests/api/sts/test_sts_keycloak_integration.js

- [ ] Doc added/updated
- [X] Tests added

## Summary by CodeRabbit

- **New Features**
  - Expanded STS `AssumeRoleWithWebIdentity` support for federated OIDC providers, including enhanced trust-policy condition evaluation and issuer-to-provider mapping.

- **Bug Fixes**
  - Strengthened IAM trust-policy validation and federated principal format checks.
  - Improved handling of `NotPrincipal` and multiple tag/principal-tag conditions with proper AND semantics.

- **Tests**
  - Added comprehensive mocked Keycloak/OIDC coverage for STS, ABAC, session tokens, and token edge cases.
  - Updated STS tests to use the S3 session-token client helper consistently.

### CODE REVIEW

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 1]
```diff
@@ -0,0 +1,2351 @@
+/* Copyright (C) 2026 NooBaa */
+/* eslint max-lines-per-function: ['error', 700] */
+/* eslint max-lines: ["error", 2600] */
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: Prefer shrinking the suite instead of raising max-lines to 2600.

I think we can use the helper functions and have less redundant code and also we can try make it more efficient.

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 2]
```diff
@@ -0,0 +1,2351 @@
...
+ * Test groups
+ * -----------
+ * A. Role creation and basic AssumeRoleWithWebIdentity flow
+ * B. Trust-policy condition validation (aws:RequestTag, aud, sub, ForAnyValue)
+ * C. S3 request flow using the AWS_SESSION_TOKEN returned by AssumeRoleWithWebIdentity
+ * D. IAM role policy enforcement with temporary credentials + aws:PrincipalTag
+ */
+
+// setup coretest first to prepare the env
+const { require_coretest, is_nc_coretest, generate_iam_client } = require('../../../system_tests/test_utils');
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: Import/use generate_sts_client / generate_s3_client (and err_code) like `test_sts.js` drop aws-sdk + hand-rolled HTTPS agent.

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 3]
```diff
@@ -0,0 +1,2351 @@
...
+            is_configured_stub.restore();
+        },
+    };
+}
+
+/**
+ * Parse the raw STS XML body out of an AWS SDK "complete" event and return
+ * the JS object produced by `http_utils.parse_xml_to_js`.
+ */
+async function assume_role_with_wi_and_parse_xml(anon_sts, params) {
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: Prefer SDK v3 `AssumeRoleWithWebIdentityCommand` and drop XML parse/[0] walking.

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 4]
```diff
@@ -0,0 +1,2351 @@
...
+async function assert_throws_async(promise, expected_code, expected_message) {
+    try {
+        await promise;
+        assert.fail(`Expected rejection with code=${expected_code} but promise resolved`);
+    } catch (err) {
+        const code = err.code || err.rpc_code;
+        if (code !== expected_code) throw err;
+        if (expected_message !== undefined && err.message !== expected_message) throw err;
+    }
+}
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js:250]: Reuse shared `assert_throws_async` / `err_code` pattern from `test_sts.js`.

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 5]
```diff
@@ -0,0 +1,2351 @@
...
+
+// ---------------------------------------------------------------------------
+// C. S3 request flow using AWS_SESSION_TOKEN
+// ---------------------------------------------------------------------------
+
+mocha.describe('Keycloak AssumeRoleWithWebIdentity - S3 with temporary credentials', function() {
+    const { rpc_client } = coretest;
+
+    const role_owner_email = 'kc-role-owner-c1';
+    const test_bucket = 'first.bucket'; // pre-existing bucket created by coretest
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: Shared `first.bucket` policy overwrite — use a dedicated bucket, or restore/delete policy in after to not affect other tests.

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 6]
```diff
@@ -0,0 +1,2351 @@
...
+        });
+
+        const creds = json.AssumeRoleWithWebIdentityResponse.AssumeRoleWithWebIdentityResult[0].Credentials[0];
+        const s3 = new AWS.S3({
+            ...sts_creds,
+            endpoint: coretest.get_https_address(),
+            accessKeyId: creds.AccessKeyId[0],
+            secretAccessKey: creds.SecretAccessKey[0],
+            sessionToken: creds.SessionToken[0],
+        });
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: Use generate_s3_client(..., session_token) instead of new AWS.S3.

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 7]
```diff
@@ -0,0 +1,2351 @@
...
+
+        const creds = json.AssumeRoleWithWebIdentityResponse.AssumeRoleWithWebIdentityResult[0].Credentials[0];
+        const s3 = new AWS.S3({
+            ...sts_creds,
+            endpoint: coretest.get_https_address(),
+            accessKeyId: creds.AccessKeyId[0],
+            secretAccessKey: creds.SecretAccessKey[0],
+            // intentionally no sessionToken
+        });
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: same here, please use shared `generate_s3_client()`

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 8]
```diff
@@ -0,0 +1,2351 @@
...
+        const creds2 = json2.AssumeRoleWithWebIdentityResponse.AssumeRoleWithWebIdentityResult[0].Credentials[0];
+
+        // Mix: access-key from session-1 with session-token from session-2
+        const s3 = new AWS.S3({
+            ...sts_creds,
+            endpoint: coretest.get_https_address(),
+            accessKeyId: creds1.AccessKeyId[0],
+            secretAccessKey: creds1.SecretAccessKey[0],
+            sessionToken: creds2.SessionToken[0],
+        });
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: please use shared `generate_s3_client()` and in other remaining places.

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 9]
```diff
@@ -0,0 +1,2373 @@
...
+            secretAccessKey: creds.SecretAccessKey[0],
+            sessionToken: expired_session_token,
+        });
+
+        await assert_throws_async(
+            s3.listObjects({ Bucket: test_bucket }).promise(),
+            S3Error.ExpiredToken.code,
+            S3Error.ExpiredToken.message
+        );
+    }, 5000);
```
[the remark this answers is not in the document]
Comment (aayushchouhan09) [reply src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: Agreed.

File: src/util/string_utils.js [hunk 10]
```diff
@@ -22,6 +22,9 @@ const AWS_POLICY_DOCUMENT_REGEXP = /^[\u0009\u000A\u000D\u0020-\u00FF]+$/;
 const AWS_ROLE_DESCRIPTION_REGEXP = /^[\u0009\u000A\u000D\u0020-\u007E\u00A1-\u00FF]*$/;
 const AWS_POLICY_SID_REGEXP = /^[A-Za-z0-9]*$/;
 const AWS_IAM_ARN_REGEXP = /^arn:aws:iam::\w{10,}:(?:root|user\/[\w\-\.\/]+)$/;
+// Matches a Federated OIDC-provider ARN used in Principal.Federated of a trust policy, e.g.:
+// arn:aws:iam::<account-id>:oidc-provider/keycloak.noobaa.svc.cluster.local:8080/realms/noobaa
+const AWS_OIDC_PROVIDER_ARN_REGEXP = /^arn:aws:iam::\w{10,}:oidc-provider\/.+$/;
```
[the remark this answers is not in the document]
Comment (aayushchouhan09) [reply src/util/string_utils.js]: Not sure if we need to put 12 or 10, Can you please verify it once and make it aligned in other files also.

**Review by aayushchouhan09 — COMMENTED** [review]
This account/STS/IAM setup is repeated in every describe. Can we have one fixture helper to cut size and keep NooBaa test style (like in `test_sts.js` and other files)?

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 11]
```diff
@@ -0,0 +1,2275 @@
...
+        self.timeout(60000);
+
+        const account_defaults = { has_login: false, s3_access: true };
+        if (is_nc_coretest) {
+            account_defaults.nsfs_account_config = {
+                uid: process.getuid(),
+                gid: process.getgid(),
+                new_buckets_path: coretest.NC_CORETEST_STORAGE_PATH,
+            };
+        }
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: Setup still duplicated across A–G (`account_defaults / account create / IAM / STS`). It would trim the file if we use helper functions for duplication code. 
I think we can still make the file more efficient and less duplicate code.

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 12]
```diff
@@ -0,0 +1,2275 @@
...
+        role_owner_account = await rpc_client.account.read_account({ email: role_owner_email });
+
+        const role_owner_keys = role_owner_account.access_keys;
+        iam_client_role_owner = generate_iam_client(
+            role_owner_keys[0].access_key.unwrap(),
+            role_owner_keys[0].secret_key.unwrap(),
+            coretest.get_https_address_iam()
+        );
+
+        anon_sts = generate_sts_client(undefined, undefined, coretest.get_https_address_sts());
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]:
```suggestion
        anon_sts = generate_sts_client('', '', coretest.get_https_address_sts());
```
like we use it in `test_sts.js`

Comment (aayushchouhan09) [reply src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: Please update it in other places as well

File: src/test/integration_tests/api/sts/test_sts_keycloak_integration.js [hunk 13]
```diff
@@ -0,0 +1,2275 @@
...
+            creds.SecretAccessKey,
+            coretest.get_http_address(),
+            creds.SessionToken
+        );
+
+        const result = await s3.listObjects({ Bucket: test_bucket });
+        assert.ok(result, 'listObjects should succeed');
+    });
+
+    mocha.it('C.2 - S3 listObjects fails without session token (temp key alone is invalid)', async function() {
```
Comment (aayushchouhan09) [inline src/test/integration_tests/api/sts/test_sts_keycloak_integration.js]: C.2–C.5 largely duplicate session-token coverage in `test_sts.js`. Keep C.1 (Keycloak S3 happy path); drop or heavily trim the rest. 
Please verify the already present tests in `test_sts.js` and remove them accordingly.

File: src/endpoint/iam/iam_utils.js [hunk 14]
```diff
@@ -1112,6 +1114,52 @@ function _validate_policy_document_iam_structure(policy_document) {
...
 
+/**
+ * _validate_assume_role_policy_document_iam_structure validates the structural correctness of an
+ * assume-role trust policy document (which is allowed to have Principal / NotPrincipal).
+ *
+ * Specifically:
+ *  - Version must be '2012-10-17' or '2008-10-17'.
+ *  - Statement must be a non-empty array.
+ *  - Principal.Federated entries, when present, must conform to the OIDC-provider ARN format:
+ *      arn:aws:iam::<12-digit-account-id>:oidc-provider/<provider-url>
```
Comment (aayushchouhan09) [inline src/endpoint/iam/iam_utils.js]: comment say 12-digit-account-id, but regex allows optional `\w+`. Align wording with actual regex.

**Review by aayushchouhan09 — COMMENTED** [review]
some comments
I think we can still make the file efficient and less duplicate code.

**Review by aayushchouhan09 — APPROVED** [review]
LGTM
